### PR TITLE
Support list (cons)

### PR DESCRIPTION
**Pull request by kmizu**
- The form of cons is `[Head|Tail]` (`Head` and `Tail` are variables of Erlang).
- Add list type (`TyList t`)
- Implement the derivation rule
- Implement the subtype relation
- Add some unit tests
- Postpone the support of nil pattern

### CODE REVIEW

**Comment by kmizu** [comment]
Remaining tasks:
- Add blackbox tests
- Fix derivation rule docs

**Comment by kmizu** [comment]
Could you review this PR ? to : @amutake and/or @yoshihiro503

File: lib/derivation.ml [hunk 1]
```diff
@@ -162,9 +162,23 @@ let rec derive context = function
...
+    | (TyUnion elements, _) -> TyUnion elements
+    | (_, TyUnion elements) -> TyUnion elements
+    | (TyAny, TyAny) -> TyAny
+    | (TyAny, TyBottom) -> TyAny
+    | (TyBottom, TyBottom) -> TyBottom
+    | (TyBottom, TyAny) -> TyAny
+    in
+    derive context hd >>= fun (ty_hd, c_hd) ->
+    derive context tl >>= fun (ty_tl, c_tl) ->
+      Ok (Type.of_elem (TyList (union_of alpha ty_hd)), C.Conj [
```
Comment (yoshihiro503) [inline lib/derivation.ml]: I think it would be better to use `sup` instead of `union_of`.

```suggestion
      Ok (Type.of_elem (TyList (Type.sup alpha ty_hd)), C.Conj [
```

File: test/unit-test/test_derivation.ml [hunk 2]
```diff
@@ -51,8 +51,7 @@ let%expect_test "derivation" =
    *   X when true -> X
    * end
    *)
-  print Context.empty (Case (Constant (-1, Number 42), [(PatVar "X", Constant (-1, Atom "true")), Var (-1, "X")]));
-  [%expect {|
+  print Context.empty (Case (Constant (-1, Number 42), [(PatVar "X", Constant (-1, Atom "true")), Var (-1, "X")])); [%expect {|
```
Comment (yoshihiro503) [inline test/unit-test/test_derivation.ml]: Is this an unnecessary change?

**Comment by amutake** [comment]
Sorry, could you rebase and change `docs/derivation-rules.saty` instead of `docs/derivation-rules.md` because `derivation-rules.md` is deleted via #207 ?
If you are not familiar with SATySFi, please let me know. I'll support or add commits instead of you.

File: test/unit-test/test_derivation.ml (hunk 2 above)
Comment (kmizu) [reply test/unit-test/test_derivation.ml]: I see. I've fixed it

**Comment by kmizu** [comment]
> Sorry, could you rebase and change `docs/derivation-rules.saty` instead of `docs/derivation-rules.md` because `derivation-rules.md` is deleted via #207 ?
> If you are not familiar with SATySFi, please let me know. I'll support or add commits instead of you.

I changed `docs/derivation-rules.saty`

**Comment by kmizu** [comment]
@yoshihiro503 Could you re-review it and do LGTM if there is no problem ?

File: docs/derivation-rules.saty [hunk 3]
```diff
@@ -153,6 +153,36 @@ but extended by remote call, local call, list, etc.
         }
       );
     >
+    +small-block<
```
Comment (amutake) [inline docs/derivation-rules.saty]: Could you add `[e | e] | []` into BNF?

Comment (kmizu) [reply docs/derivation-rules.saty]: OK. I'll add it.

Comment (kmizu) [reply docs/derivation-rules.saty]: I added it in https://github.com/dwango/fialyzer/pull/191/commits/aa4740d501a0f73ac3e2ef6b0fed7771f0180f6c
